### PR TITLE
Return to correct file position after power loss

### PR DESCRIPTION
**Pull request by espr14**
### Description
`Planner::block_buffer[Planner::block_buffer_tail]` is the current block but printer is printing the previous block. This version resumes from the correct file position during power loss recovery.

### Requirements
Power loss feature (pin, trigger).

### Benefits
More precise return from power loss.

### Configurations
`POWER_LOSS_RECOVERY` must be enabled and set according to printer.

### CODE REVIEW

File: Marlin/src/module/stepper.cpp [hunk 1]
```diff
@@ -2019,7 +2019,7 @@ uint32_t Stepper::block_phase_isr() {
         cutter.apply_power(current_block->cutter_power);
       #endif
 
-      TERN_(POWER_LOSS_RECOVERY, recovery.info.sdpos = current_block->sdpos);
+      TERN_(POWER_LOSS_RECOVERY, recovery.info.sdpos = Planner::block_buffer[BLOCK_MOD(Planner::block_buffer_tail - 1)].sdpos);
```
Comment (thinkyhead) [inline Marlin/src/module/stepper.cpp:2022]: At this point the previous block is finished and we are about to start a new `current_block`, so we *do* want the SD position from the `current_block`. It *should* contain the position of the corresponding command in the file. If we're not getting the right value here into `recovery.info.sdpos` then we should look at the code which sets the `sdpos` in the block structure instead.

**Comment by thinkyhead** [comment]
Here is the intended procedure:

- `recovery.cmd_sdpos` contains the index of the command just fetched from the SD Card.

- The value of `recovery.cmd_sdpos` is managed by `get_sdcard_commands` ...
  - Call `commit_command` to copy `recovery.cmd_sdpos` into `recovery.sdpos[index_w]`.
  - Set `recovery.cmd_sdpos` to the `card.sdpos`, which is the start of the _next_ command.

- Upon command execution `process_next_command` sets `recovery.queue_index_r` to the ring queue index of the current running G-code.

- In `Planner::_populate_block` set the block's `sdpos` from `recovery.command_sdpos()` (`recovery.sdpos[recovery.queue_index_r]`).

- When the Stepper ISR gets a new block it copies the `sdpos` from the block to `recovery.info.sdpos` so PLR is ready to go.

So, we'll need to examine each of these steps to see that it's doing the right thing.

**Comment by CRCinAU** [comment]
In a recent build (20210531), I've had someone tell me that after the prompt to resume, the print started from line 1 of the G-Code again.

I'm trying to get more information, but it may well be that something isn't quite right.

I'll add more comments if I can come up with anything more concrete...

**Comment by Rick615** [comment]
I just installed (20210531), and I can confirm that after the prompt to resume, the print does start from the beginning again and prints on top of what was already there if you did not remove it.  It does write to the SD card, because there is an extra file there.

**Comment by thinkyhead** [comment]
The `DEBUG_POWER_LOSS_RECOVERY` option can be useful here. With that option enabled there will be some extra log output from all power-loss methods. And you can use these commands to test the system and see what's going on:
- `M413 D` : Display the contents of the PLR file.
- `M413 E` : Report if the PLR file exists.
- `M413 V` : Report whether the PLR file is valid.
- `M413 W` : Write the PLR file.
- `M413 P` : Purge the PLR file.
- `M413 O` : Simulate a power outage.

**Comment by thinkyhead** [comment]
Were you able to use `DEBUG_POWER_LOSS_RECOVERY` to gather more information? Please post the logs.

**Comment by espr14** [comment]
My mistake. It's done right. If you want to solve the "print continues from the beginning" it would be better to start a new issue.

**Pull request closed without merging**